src/main/java/org/zalando/nakadi/service/CursorOperationsService.java

```diff
@@ -0,0 +1,209 @@
+package org.zalando.nakadi.service;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.zalando.nakadi.domain.NakadiCursorLag;
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.NakadiCursorDistanceQuery;
+import org.zalando.nakadi.domain.NakadiCursorDistanceResult;
+import org.zalando.nakadi.domain.PartitionStatistics;
+import org.zalando.nakadi.domain.ShiftedNakadiCursor;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation;
+import org.zalando.nakadi.exceptions.runtime.MyNakadiRuntimeException1;
+import org.zalando.nakadi.repository.TopicRepository;
+import org.zalando.nakadi.service.timeline.TimelineService;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.CURSORS_WITH_DIFFERENT_PARTITION;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_OFFSET_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_TIMELINE_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.TIMELINE_NOT_FOUND;
+
+@Service
+public class CursorOperationsService {
+    private final TimelineService timelineService;
+
+    @Autowired
+    public CursorOperationsService(final TimelineService timelineService) {
+        this.timelineService = timelineService;
+    }
+
+    public List<NakadiCursorDistanceResult> calculateDistance(final List<NakadiCursorDistanceQuery> cursors)
+        throws InvalidCursorOperation {
+        return cursors.stream().map(this::calculateDistance).collect(Collectors.toList());
+    }
+
+    public NakadiCursorDistanceResult calculateDistance(final NakadiCursorDistanceQuery query)
+            throws InvalidCursorOperation {
+        final NakadiCursor initialCursor = query.getInitialCursor();
+        final NakadiCursor finalCursor = query.getFinalCursor();
+
+        // Validate query
+        if (!initialCursor.getPartition().equals(finalCursor.getPartition())) {
+            throw new InvalidCursorOperation(CURSORS_WITH_DIFFERENT_PARTITION);
+        } else if (initialCursor.getTimeline().getOrder() > finalCursor.getTimeline().getOrder()) {
+            throw new InvalidCursorOperation(INVERTED_TIMELINE_ORDER);
```

Comment:
but, in the other had, the word `distance` implies something that is always positive. So it would be the modulus of that? To avoid confusion when interpreting results, I thought about making it more strict, with less options. If `initial` is always older than `final` than `distance` is always positive and very easy to understand.